src/main/java/org/candlepin/resource/util/EntitlementEnvironmentFilter.java

```diff
@@ -0,0 +1,147 @@
+ * implied, including the implied warranties of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. You should have received a copy of GPLv2
+ * along with this software; if not, see
+ * http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt.
+ *
+ * Red Hat trademarks are not licensed under GPLv2. No permission is
+ * granted to use or replicate Red Hat trademarks that are incorporated
+ * in this software or its documentation.
+ */
+package org.candlepin.resource.util;
+
+import org.candlepin.model.Consumer;
+import org.candlepin.model.Entitlement;
+import org.candlepin.model.EntitlementCurator;
+import org.candlepin.model.EnvironmentCurator;
+
+import org.apache.commons.collections.CollectionUtils;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+/*
+ * The EntitlementEnvironmentFilter class filters the consumers entitlement who's entitlement certificates are
+ * needed to be re-generated because of changes (addition, removal or re-prioritization) in
+ * consumer existing environment. This filtering avoids re-generating all of the
+ * consumer entitlement certificates.
+ */
+public class EntitlementEnvironmentFilter {
+    private EntitlementCurator entitlementCurator;
+    private EnvironmentCurator environmentCurator;
+    private List<String> preExistingEnvs;
+    private List<String> updatedEnvs;
+    private Consumer consumer;
+
+    public EntitlementEnvironmentFilter(EntitlementCurator entitlementCurator,
+        EnvironmentCurator environmentCurator) {
+        this.entitlementCurator = entitlementCurator;
+        this.environmentCurator = environmentCurator;
+    }
+
+    public EntitlementEnvironmentFilter setPreExistingEnvironments(List<String> preExistingEnvs) {
+        this.preExistingEnvs = preExistingEnvs;
+        return this;
+    }
+
+    public EntitlementEnvironmentFilter setUpdatedEnvironment(List<String> updatedEnvs) {
+        this.updatedEnvs = updatedEnvs;
+        return this;
+    }
+
+    public EntitlementEnvironmentFilter setConsumerToBeUpdated(Consumer consumer) {
+        this.consumer = consumer;
+        return this;
+    }
+
+    /**
+     * Returns the list of environment IDs which got added, removed or re-prioritized
+     * from consumers current existing environments.
+     *
+     * @return
+     *  Returns the unique environmentIds which got added, removed or re-prioritized
+     */
+    private List<String> getEnvironmentChangeList() {
+        Set<String> envChangeList = new HashSet<>();
+
+        // Envs Added
+        envChangeList.addAll(CollectionUtils.subtract(this.updatedEnvs, this.preExistingEnvs));
+
+        // Envs removed
+        envChangeList.addAll(CollectionUtils.subtract(this.preExistingEnvs, this.updatedEnvs));
+
+        // Envs reordered (Index within list changes)
+        for (String env : this.updatedEnvs) {
+            if (this.preExistingEnvs.indexOf(env) != this.updatedEnvs.indexOf(env)) {
```

Comment:
This is not exactly correct. For example:
- if `preExistingEnvs: A B C` 
- and `updatedEnvs: B C`
- Then this code will consider both B and C to be re-ordered, while they haven't been re-ordered compared to each other.
Another example
- if `preExistingEnvs: A B` 
- and `updatedEnvs: A C B`
- Same here: A and B are not reordered (compared to each other).

The scenario we want to catch here is something like this:
- if `preExistingEnvs: A B` 
- and `updatedEnvs: B C A`
- Now A and B are reordered (compared to each other).

Reply:
Fixed